docs/rfcs/0003-sql-datastore.md

```diff
@@ -0,0 +1,281 @@
+
+```sql
+# PostgreSQL
+CREATE TABLE projects (
+  id UUID PRIMARY KEY,
+  data JSONB NOT NULL,
+  disabled BOOL NOT NULL,
+  created_at BIGINT NOT NULL,
+  updated_at BIGINT NOT NULL
+);
+CREATE TABLE applications (
+  id UUID PRIMARY KEY,
+  project_id UUID NOT NULL,
+  data JSONB NOT NULL,
+  disabled BOOL NOT NULL,
+  created_at BIGINT NOT NULL,
+  updated_at BIGINT NOT NULL
+);
+
+# MySQL
+CREATE TABLE projects (
+  id BINARY(16) PRIMARY KEY,
+  data JSON NOT NULL,
+  disabled BOOL NOT NULL,
+  created_at INT(11) NOT NULL,
+  updated_at INT(11) NOT NULL
+) ENGINE=InnoDB;
+
+CREATE TABLE applications (
+  id BINARY(16) PRIMARY KEY,
+  project_id BINARY(16) NOT NULL,
+  data JSON NOT NULL,
+  disabled BOOL NOT NULL,
+  created_at INT(11) NOT NULL,
+  updated_at INT(11) NOT NULL
+) ENGINE=InnoDB;
+```
+
+Note:
+- Both Jsonb and Json point to the type of storing JSON data in those databases. The key difference is that Json data is stored as an exact copy of the JSON input text, whereas Jsonb stores data in a decomposed binary form. Jsonb seems a better choice since it supports indexing with significant performance increases.
+- Save those data redundantly (both as table columns and as attributes of the JSON data) to reduce model object initialization costs.
+# Alternatives
+
+Currently, we consider between MySQL and PostgreSQL for those support for NoSQL features, PostgreSQL has a longer time in this field. Some considering factors:
+- ability to index a specific attribute of the JSON field
+- supported operators on specific JSON field
+- performance of query operation on indexed JSON field
+- able to keep the advantage of schemaless pattern
+
+## Ability to index a specific attribute of the JSON field
+
+We have 2 points which have to be considered.
+
+1. UUID as primary keys should not affect queries performance
+
+While MySQL does not have UUID data type, since we're using `application side UUID generate` pattern, (which mean to MySQL, those ids are true random UUID) store those ids under `VARCHAR(32)` data type is costly for both read and write operation due to the indexes does not work, using `UUID_TO_BIN` & `BIN_TO_UUID` with `swap_flag` and store data under `BINARY(16)` would help.
+
+From MySQL docs
+```
+  o If swap_flag is 0, the two-argument form is equivalent to the
+    one-argument form. The binary result is in the same order as the
+    string argument.
+
+  o If swap_flag is 1, the format of the return value differs: The
+    time-low and time-high parts (the first and third groups of
+    hexadecimal digits, respectively) are swapped. This moves the more
+    rapidly varying part to the right and can improve indexing
+    efficiency if the result is stored in an indexed column.
+
+```
+
+PostgreSQL has UUID type to store this kind of data and support it as primary key as well.
+
+MySQL: 👍 PostgreSQL: 👍
```

Comment:
I vote for PostgreSQL to have 👍 👍  because of its native UUID type.

Reply:
sure 👍 will do 🙆‍♂️